### PR TITLE
Arp/api

### PR DESCRIPTION
**Pull request by alyssa**

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -112,3 +129,303 @@ func (api *API) statsHandler(ctx router.HTTPContext) {
...
+
+	// Convert any float64 values in entity to int
+	for k, v := range entity {
+		// Continue if v is not type float64
+		if reflect.TypeOf(v).Kind() != reflect.Float64 {
+			continue
+		}
+
+		// Trunc returns the integer value of x.
+		if v == math.Trunc(v.(float64)) {
```
[the remark this answers is not in the document]
Comment (alyssa) [reply api/api.go]: we don't use `int(x)` in the check as `v` is type `float64` so we cannot compare it to an int.

i'll add a code comment to clarify the intention

File: api/api.go [hunk 2]
```diff
@@ -112,3 +129,303 @@ func (api *API) statsHandler(ctx router.HTTPContext) {
...
+		} else {
+			ctx.APIError(router.ErrInternal, "Uknown error when creating entity(error: %s)", err)
+			return
+		}
+	}
+
+	if out, err := marshal(ids[0]); err != nil {
+		ctx.APIError(router.ErrInternal, "Can't encode response (error: %s)", err)
+	} else {
+		fmt.Fprintln(ctx.Response, string(out))
```
[the remark this answers is not in the document]
Comment (alyssa) [reply api/api.go]: ah ok good point. i've updated that

File: api/api.go [hunk 3]
```diff
@@ -112,3 +129,303 @@ func (api *API) statsHandler(ctx router.HTTPContext) {
...
+				Operator: "=",
+				Value:    id,
+			},
+		},
+	}
+}
+
+// marshal is a helper function to nicely print JSON.
+func marshal(v interface{}) ([]byte, error) {
+	return json.MarshalIndent(v, "", "  ")
```
[the remark this answers is not in the document]
Comment (alyssa) [reply api/api.go]: good point and true. thanks for pointing it out. i updated it to just do `json.Marshal`

File: api/api_test.go [hunk 4]
```diff
@@ -1,3 +1,372 @@
...
+	"net/http/httptest"
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/go-test/deep"
+	"github.com/square/etre/db"
+	"github.com/square/etre/query"
+	"github.com/square/etre/router"
+	testutil "github.com/square/etre/test"
```
[the remark this answers is not in the document]
Comment (alyssa) [reply api/api_test.go]: ok i removed the rename

**Comment by alyssa** [comment]
@daniel-nichter I've addressed your feedback, completed writing out the tests, and made it so the query is in the url now (`/api/v1?query="foo=bar,x>2"`). PTAL. Thanks!

File: api/api.go [hunk 5]
```diff
@@ -112,3 +125,342 @@ func (api *API) statsHandler(ctx router.HTTPContext) {
...
+	ConvertFloat64ToInt(entity)
+
+	for _, v := range entity {
+		if !validValueType(v) {
+			ctx.APIError(router.ErrBadRequest, "Invalid Value Type: %v. Value type must be string or int", reflect.TypeOf(v))
+		}
+	}
+
+	entities := []db.Entity{entity}
+	ids, err := c.CreateEntities(entities)
```
Comment (daniel-nichter) [inline api/api.go]: Since `entities` isn't used elsewhere, you can just `c.CreateEntities([]db.Entity{entity})`.

File: api/api.go [hunk 6]
```diff
@@ -112,3 +125,342 @@ func (api *API) statsHandler(ctx router.HTTPContext) {
...
+	if err != nil {
+		if _, ok := err.(db.ErrCreate); ok {
+			ctx.APIError(router.ErrInternal, "Error creating entity (error: %s)", err)
```
Comment (daniel-nichter) [inline api/api.go]: `"Error creating entity: %s"`. The error is not a parenthetical note to the message, it _is_ the message. :-)

File: api/api.go [hunk 7]
```diff
@@ -112,3 +125,342 @@ func (api *API) statsHandler(ctx router.HTTPContext) {
...
+		ctx.APIError(router.ErrInternal, "Can't encode response (error: %s)", err)
+		return
+	}
+
+	fmt.Fprintln(ctx.Response, string(out))
+}
+
+func getEntityHandler(ctx router.HTTPContext, c db.Connector) {
+	if len(ctx.Arguments) != 2 {
+		ctx.APIError(router.ErrMissingParam, "Missing param: id")
```
Comment (daniel-nichter) [inline api/api.go]: What's the 1st arg? Is it possible we'd have 3 and so id isn't missing, we just have more than expected? I guess what strikes me is that len of some args does not seem directly or obviously related to a param missing. No big deal, I just want to make sure error messages are good/accurate.

File: api/api.go [hunk 8]
```diff
@@ -112,3 +125,342 @@ func (api *API) statsHandler(ctx router.HTTPContext) {
...
+		ctx.APIError(router.ErrInternal, "Can't encode response (error: %s)", err)
+		return
+	}
+
+	fmt.Fprintln(ctx.Response, string(out))
+}
+
+func getEntitiesHandler(ctx router.HTTPContext, c db.Connector) {
+	// Translate URL query to query struct
+	requestLabelSelector := ctx.Request.Form["query"][0]
```
Comment (daniel-nichter) [inline api/api.go]: If `"query"` is missing, then `[0]` might cause a panic trying to access nil.

File: api/api.go [hunk 9]
```diff
@@ -112,3 +125,342 @@ func (api *API) statsHandler(ctx router.HTTPContext) {
...
+	if requestLabelSelector == "" {
+		ctx.APIError(router.ErrMissingParam, "Missing param: query")
+		return
+	}
+	q, err := query.Translate(requestLabelSelector)
+	if err != nil {
+		ctx.APIError(router.ErrInternal, "Can't translate query (error: %s)", err)
```
Comment (daniel-nichter) [inline api/api.go]: Query translation is an internal matter. As user, if I saw this error message I wouldn't understand (I'm not aware of any translations). Maybe something more generic like `"Invalid query: %s", err`, where `err` is hopefully specific about the problem.

File: api/api.go [hunk 10]
```diff
@@ -112,3 +125,342 @@ func (api *API) statsHandler(ctx router.HTTPContext) {
...
+	if err != nil {
+		ctx.APIError(router.ErrInternal, "Can't decode request body (error: %s)", err)
+		return
+	}
+
+	ConvertFloat64ToInt(entity)
+
+	for _, v := range entity {
+		if !validValueType(v) {
+			ctx.APIError(router.ErrBadRequest, "Invalid Value Type: %v. Value type must be string or int", reflect.TypeOf(v))
```
[the remark this answers is not in the document]
Comment (alyssa) [reply api/api.go]: It says which label but not which key. I've updated to include the key, value, and type of value.

File: api/api.go (hunk 6 above)
Comment (alyssa) [reply api/api.go]: heh good point...updated for this and all other ones like it

File: api/api.go (hunk 7 above)
Comment (alyssa) [reply api/api.go]: The first arg is the path (such as `/api/v1/entity`). Given how the route is defined for this particular handler, there's no possibility to have a 3rd.

```
api.Router.AddRoute(API_ROOT+"entity", api.entityHandler, "api-entity")
api.Router.AddRoute(API_ROOT+"entity/"+REQUEST_ID_PATTERN, api.entityHandler, "api-entity")
```

File: api/api.go (hunk 8 above)
Comment (alyssa) [reply api/api.go]: ah good catch. updated the code + added test cases.

File: api/api.go (hunk 9 above)
Comment (alyssa) [reply api/api.go]: good point in translate being internal. i updated the messaging. the error message ultimately comes from KLS code i think which is pretty specific to the error (usually wrong type or something)

**Pull request merged**